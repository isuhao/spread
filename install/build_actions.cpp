#include "build_actions.hpp"

#include "rules/install_finder.hpp"
#include <stdexcept>
#include <boost/filesystem.hpp>

using namespace Spread;

//#define DEBUG_PRINT
#ifdef DEBUG_PRINT
#include <iostream>
#define PRINT(a) std::cout << __LINE__ << ": " << a << "\n";

static void print(ActionMap &output)
{
  using namespace std;
  cout << "Got " << output.size() << " actions:\n";
  ActionMap::iterator it;
  for(it = output.begin(); it != output.end(); it++)
    {
      Action &a = it->second;

      cout << it->first << " : ";

      if(a.isNone())
        {
          cout << "UNHANDLED\n";
        }
      else if(a.isRule())
        {
          const Rule &r = *a.rule;

          cout << r.ruleString << endl;

          // Print rule dependencies
          for(int i=0; i<r.deps.size(); i++)
            cout << "  <= " << r.deps[i] << endl;
        }
      else if(a.isCopy())
        {
          cout << "COPY from " << a.source << endl;
        }
      else assert(0);

      // Print output files
      std::set<std::string>::iterator it2;
      for(it2 = a.destlist.begin(); it2 != a.destlist.end(); it2++)
        cout << "  => " << *it2 << endl;
    }
}
#else
#define PRINT(a)
#endif

void ActionBuilder::addHint(const Hash &hint)
{
  // Find the matching archive
  const ArcRuleData *data = rules.findArchive(hint);
  if(!data) return;

  // Load the directory object
  DirectoryCPtr dir = cache.loadDir(data->dirHash);
  if(!dir) return;

  // Prime the ArcRuleSet with the data
  arcs.addArchive(data->arcHash, dir, data->ruleString);
}

void ActionBuilder::addDep(const std::string &file, const Hash &hash)
{
  list.dir[file] = hash;
}

void ActionBuilder::addDir(const Directory *dir)
{
  Directory::DirMap::const_iterator it;
  for(it = dir->dir.begin(); it != dir->dir.end(); it++)
    addDep(it->first, it->second);
}

void ActionBuilder::addDir(const Hash &search, bool alsoAsHint)
{
  PRINT("addDir(hash=" << hash << ", alsoAsHint=" << alsoAsHint << ")");

  // Check if this is an archive hash, if so, pick out the dir hash.
  Hash dirHash = search;
  const ArcRuleData *data = rules.findArchive(search);
  if(data)
    dirHash = data->dirHash;

  DirectoryCPtr dir = cache.loadDir(dirHash);

  if(!dir)
    {
      PRINT("  No dir found, going blind.");

      // Add blind. Requires a matching archive rule.
      if(!data)
        throw std::runtime_error("Hash " + search.toString() + " matches neither a directory file nor any archive rule.");

      PRINT("  Adding archive " << data->arcHash);
      blind.push_back(data->arcHash);
      return;
    }

  PRINT("  Found directory with " << dir->dir.size() << " elements");

  addDir(dir);

  // Add the hint if the user requested it
  if(alsoAsHint && data)
    arcs.addArchive(data->arcHash, dir, data->ruleString);
}

void ActionBuilder::build(ActionMap &output)
{
  PRINT("build()");

  InstallFinder finder(arcs, cache.index);

  namespace bs = boost::filesystem;

  PRINT("DEPLIST");

  // Build dependency list
  {
    bs::path pdir = prefix;
    Directory::DirMap::const_iterator it;
    for(it = list.dir.begin(); it != list.dir.end(); it++)
      {
        // Add path prefix, if any
        bs::path file;
        if(prefix != "")
          file = pdir/it->first;
        else
          file = it->first;

        finder.addDep(file.string(), it->second);
      }
  }

  // Add "blind" archives
  for(int i=0; i<blind.size(); i++)
    finder.addBlind(prefix, blind[i]);

  PRINT("PERFORMING");

  bool res = finder.perform(output);

#ifdef DEBUG_PRINT
  print(output);
#endif

  if(!res)
    {
      // Build an error message
      std::string err = "Don't know how to satisfy the following dependencies:";

      // Loop through output and add the names of all the missing files
      ActionMap::const_iterator it;
      for(it = output.begin(); it != output.end(); it++)
        {
          const Action &a = it->second;
          if(a.isNone())
            {
              err += "\n  " + it->first.toString();
              std::set<std::string>::const_iterator it2;
              for(it2 = a.destlist.begin(); it2 != a.destlist.end(); it2++)
                err += "\n    " + *it2;
            }
        }

      throw std::runtime_error(err);
    }
}
