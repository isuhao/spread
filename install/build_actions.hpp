#ifndef __SPREAD_BUILDACTIONS_HPP_
#define __SPREAD_BUILDACTIONS_HPP_

#include "rules/arcruleset.hpp"
#include "rules/ruleset.hpp"
#include "dir/directory.hpp"
#include "cache/cache.hpp"
#include "rules/actions.hpp"

/* ActionBuilder produces an ActionMap list of actions from a
   requested list of output files or directories.

   The produced ActionMap contains all the necessary information to
   complete an installation based on existing rules and data. It can
   be sent to ActionInstaller to perform the actual installation.
 */

namespace Spread
{
  struct ActionBuilder
  {
    ActionBuilder(Cache::Cache &_cache, RuleSet &_rules,
                  const std::string &_prefix = "")
      : cache(_cache), rules(_rules), arcs(&rules), prefix(_prefix)
    {}

    // Add archive hint. The hash can be either for the archive file
    // itself or for the directory of its contents.
    void addHint(const Hash &hint);

    // Add a single dependency. The file path is taken to be relative
    // to 'prefix'.
    void addDep(const std::string &file, const Hash &hash);

    // Add a directory of dependencies
    void addDir(const Directory *dir);
    void addDir(DirectoryCPtr dir) { addDir(dir.get()); }

    /* Add a directory by hash. This assumes the directory is
       available as a loadable cached object somewhere.

       If alsoAsHint==true (default), then any archive rule that
       contains this directory is also loaded.
     */
    void addDir(const Hash &hash, bool alsoAsHint = true);

    /* Build the complete install action list from the given
       dependencies.
     */
    void build(ActionMap &output);

  private:
    Directory list;

    Cache::Cache &cache;
    RuleSet &rules;
    ArcRuleSet arcs;

    // Path prefix, added to every filename in the list
    std::string prefix;
  };
}
#endif
