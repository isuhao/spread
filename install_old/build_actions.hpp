#ifndef __SPREAD_BUILDACTIONS_HPP_
#define __SPREAD_BUILDACTIONS_HPP_

#include "rules/arcruleset.hpp"
#include "rules/ruleset.hpp"
#include "dir/directory.hpp"
#include "cache/cache.hpp"
#include "rules/actions.hpp"
#include <vector>

/* ActionBuilder produces an ActionMap list of actions from a
   requested list of output files or directories.

   The produced ActionMap contains all the necessary information to
   complete an installation based on existing rules and data. It can
   be sent to ActionInstaller to perform the actual installation.
 */

namespace Spread
{
  struct ActionBuilder
  {
    ActionBuilder(Cache::Cache &_cache, RuleSet &_rules,
                  const std::string &_prefix = "")
      : cache(_cache), rules(_rules), arcs(&rules), prefix(_prefix)
    {}

    // Add archive hint. The hash can be either for the archive file
    // itself or for the directory of its contents.
    void addHint(const Hash &hint);

    // Add a single dependency. The file path is taken to be relative
    // to 'prefix'.
    void addDep(const std::string &file, const Hash &hash);

    // Add a directory of dependencies
    void addDir(const Directory *dir, const std::string &path = "");
    void addDir(DirectoryCPtr dir, const std::string &path = "")
    { addDir(dir.get(), path); }

    /* Add a directory by hash. This assumes that either the directory
       is available as a loadable cached object somewhere, or that
       there is a matching archive rule for the directory.

       If alsoAsHint==true (default), then any archive rule that
       contains this directory is loaded, if it exists.

       If the directory object does not exist, but an archive rule is
       found, the directory is added as a "blind" unpack. That means
       that whatever the archive contains is used as the output
       directory.
     */
    void addDir(const Hash &hash, bool alsoAsHint = true, const std::string &path = "");

    /* Build the complete install action list from the given
       dependencies.
     */
    void build(ActionMap &output);

  private:
    Directory list;

    Cache::Cache &cache;
    RuleSet &rules;
    ArcRuleSet arcs;

    // List of blind archive unpacks
    std::vector<Hash> blind;

    // Path prefix, added to every filename in the list
    std::string prefix;
  };
}
#endif
