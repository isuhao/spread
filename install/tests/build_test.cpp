#include "build_actions.hpp"
#include <iostream>

using namespace std;
using namespace Spread;

Hash hello("hello", 5);
Hash world("world", 5);
Hash testsh("ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ");
Hash zipfile("bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ");
Hash zipzip("RKEFhXsBxTYXmQt4oMxU_RQk_t7F0Oj8RMGALSJKh6UfAg");
Hash testzip("UXWlJa1SLdf0UcJTqiM8dKWxKOORKX_GlpRhnnT2TOlMAQ");

DirectoryPtr testDir, zipDir, zipzipDir;

Hash testDirH, zipDirH, zipzipDirH;

void setupDirs()
{
  testDir.reset(new Directory);
  zipDir.reset(new Directory);
  zipzipDir.reset(new Directory);

  testDir->dir["hello.out"] = hello;
  testDir->dir["dir/world.out"] = world;
  zipDir->dir["test.sh"] = testsh;
  zipzipDir->dir["testsh.zip"] = zipfile;

  using namespace Mangle::Stream;

  testDirH = testDir->write("_test.dir");
  zipDirH = zipDir->write("_zip.dir");
  zipzipDirH = zipzipDir->write("_zipzip.dir");
}

void print(ActionMap &output)
{
  cout << "Got " << output.size() << " actions:\n";
  ActionMap::iterator it;
  for(it = output.begin(); it != output.end(); it++)
    {
      Action &a = it->second;

      cout << it->first << " : ";

      if(a.isNone())
        {
          cout << "UNHANDLED\n";
        }
      else if(a.isRule())
        {
          const Rule &r = *a.rule;

          cout << r.ruleString << endl;

          // Print rule dependencies
          for(int i=0; i<r.deps.size(); i++)
            cout << "  <= " << r.deps[i] << endl;
        }
      else if(a.isCopy())
        {
          cout << "COPY from " << a.source << endl;
        }
      else assert(0);

      // Print output files
      std::set<std::string>::iterator it2;
      for(it2 = a.destlist.begin(); it2 != a.destlist.end(); it2++)
        cout << "  => " << *it2 << endl;
    }
}

void test(ActionBuilder &build, const std::string &msg)
{
  cout << "\n" << msg << ":\n";
  ActionMap acts;
  try { build.build(acts); }
  catch( std::exception &e )
    {
      cout << "CAUGHT: " << e.what() << endl;
      return;
    }
  print(acts);
}

#define ADD(f,h) assert(cache.index.addFile(f)==(h))

int main()
{
  setupDirs();
  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    test(build, "Empty builder");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    build.addDep("output.dat", hello);
    test(build, "Request non-existing file");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "prefix");
    build.addDep("output.dat", hello);
    ADD("hello.txt",hello);
    test(build, "Request existing file");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "prefix");
    cout << "\nAdding non-existing dir:\n";
    try { build.addDir(testDirH); }
    catch( std::exception &e )
      {
        cout << "CAUGHT: " << e.what() << endl;
      }
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "some/dir/");
    build.addDep("test.sh", testsh);
    build.addHint(zipfile);
    //rules.addArchive(zipfile, zipDirH, "ZIP1");
    test(build, "Add hint but not archive");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "some/dir/");
    build.addDep("test.sh", testsh);
    //build.addHint(zipfile);
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    test(build, "Add archive but not hint");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "some/dir/");
    build.addDep("test.sh", testsh);
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    build.addHint(zipfile);
    test(build, "Both archive and hint, but no loadable dir");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "some/dir/");
    build.addDep("test.sh", testsh);
    ADD("_zip.dir",zipDirH);
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    build.addHint(zipfile);
    test(build, "WITH loadable dir");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "some/dir/");
    build.addDep("test.sh", testsh);
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    ADD("_zip.dir",zipDirH);
    ADD("testsh.zip",zipfile);
    build.addHint(zipfile);
    test(build, "Full working case");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "out");
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    ADD("_zip.dir",zipDirH);
    ADD("testsh.zip",zipfile);
    build.addDir(zipDirH);
    test(build, "Implicit hinting through addDir");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "out");
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    ADD("_zip.dir",zipDirH);
    ADD("testsh.zip",zipfile);
    ADD("hello.txt",hello);
    build.addDir(testDir);
    build.addDir(zipDirH);
    test(build, "Three files from different sources, one missing");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "out/");
    rules.addArchive(zipfile, zipDirH, "ZIP1");
    ADD("_zip.dir",zipDirH);
    ADD("testsh.zip",zipfile);
    ADD("hello.txt",hello);
    build.addDir(testDir);
    build.addDir(zipDirH);
    rules.addURL(world, "http://world-url.com/");
    test(build, "Three files from different sources, all present");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    build.addDep("test.sh", hello);
    test(build, "Overwriting existing file with wrong data");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    build.addDep("test.sh", testsh);
    test(build, "Overwriting existing file with right data");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules, "out");
    rules.addArchive(zipfile, zipDirH, "TESTSH ZIP");
    rules.addArchive(testzip, testDirH, "TESTDIR ZIP");
    ADD("_zip.dir",zipDirH);
    ADD("_test.dir",testDirH);
    //ADD("testsh.zip",zipfile);
    rules.addURL(zipfile, "zip1URL");
    ADD("testdir.zip",testzip);
    build.addDir(testDirH);
    build.addDir(zipDirH);
    test(build, "Unzipping two zips into one dir");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    rules.addArchive(zipfile, zipDirH, "TESTSH ZIP");
    rules.addArchive(testzip, testDirH, "TESTDIR ZIP");
    ADD("_zip.dir",zipDirH);
    ADD("_test.dir",testDirH);
    //ADD("testsh.zip",zipfile);
    rules.addURL(zipfile, "zip1URL");
    ADD("testdir.zip",testzip);
    build.addHint(testDirH);
    build.addHint(zipDirH);
    test(build, "Hinting to two zips (no actual deps)");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    rules.addArchive(zipfile, zipDirH, "TESTSH ZIP");
    rules.addArchive(testzip, testDirH, "TESTDIR ZIP");
    ADD("_zip.dir",zipDirH);
    ADD("_test.dir",testDirH);
    //ADD("testsh.zip",zipfile);
    rules.addURL(zipfile, "zip1URL");
    ADD("testdir.zip",testzip);
    build.addHint(testDirH);
    build.addHint(zipDirH);
    build.addDep("file_hello", hello);
    build.addDep("file_world", world);
    build.addDep("file_testsh", testsh);
    build.addDep("file_hello_again", hello);
    test(build, "Adding stand-alone files from two zips");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    rules.addArchive(zipfile, zipDirH, "ZIP");
    rules.addArchive(zipzip, zipzipDirH, "ZIPZIP");
    ADD("_zip.dir",zipDirH);
    ADD("_zipzip.dir",zipzipDirH);
    ADD("zipzip.zip",zipzip);
    build.addHint(zipDirH);
    //build.addHint(zipzipDirH);
    build.addDep("testsh_output", testsh);
    test(build, "Zip in zip, hinting at inner zip");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    rules.addArchive(zipfile, zipDirH, "ZIP");
    rules.addArchive(zipzip, zipzipDirH, "ZIPZIP");
    ADD("_zip.dir",zipDirH);
    ADD("_zipzip.dir",zipzipDirH);
    ADD("zipzip.zip",zipzip);
    //build.addHint(zipDirH);
    build.addHint(zipzipDirH);
    build.addDep("testsh_output", testsh);
    test(build, "Zip in zip, hinting at outer zip");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);
    rules.addArchive(zipfile, zipDirH, "ZIP");
    rules.addArchive(zipzip, zipzipDirH, "ZIPZIP");
    ADD("_zip.dir",zipDirH);
    ADD("_zipzip.dir",zipzipDirH);
    ADD("zipzip.zip",zipzip);
    build.addHint(zipDirH);
    build.addHint(zipzipDirH);
    build.addDep("testsh_output", testsh);
    test(build, "Zip in zip, hinting at both zips");
  }

  {
    Cache::Cache cache;
    RuleSet rules;
    ActionBuilder build(cache, rules);

    rules.addURL(hello, "url 3 10", 3, 10);
    rules.addURL(hello, "url 4", 4);
    rules.addURL(hello, "url default");
    build.addDep("output.hello", hello);
    test(build, "All URLs present");
    rules.reportBrokenURL(hello, "url 4");
    test(build, "Disabled url 4");
    rules.reportBrokenURL(hello, "url 3 10");
    test(build, "Disabled the other url also");
    rules.reportBrokenURL(hello, "url default");
    test(build, "Disabled everything");
  }

  /* Unhandled case: 'weird' unpacking data. We don't know how, or
     even if, this will be implemented yet.

     This will possibly just be an extra hash parameter to
     addArchive(), passed to ArcRuleSet::addArchive() and stored in
     ArcRule. If it ends up being a Directory object, then we pass it
     as a DirectoryCPtr. Otherwise we can keep it as a hash and load
     it from cache when setting up UnpackHash.
   */

  cout << endl;
  return 0;
}
