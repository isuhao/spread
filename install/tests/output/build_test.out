
Empty builder:
Got 0 actions:

Request non-existing file:
CAUGHT: Don't know how to satisfy the following dependencies:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
    /home/mortennk/koding/spread/install/tests/output.dat

Request existing file:
Got 1 actions:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : COPY from /home/mortennk/koding/spread/install/tests/hello.txt
  => /home/mortennk/koding/spread/install/tests/prefix/output.dat

Adding non-existing dir:
CAUGHT: Directory not found: b0pJQX7I8U2_IpfK3fsppNT10LS9Q09C2FchwEYXuJFy

Add hint but not archive:
CAUGHT: Don't know how to satisfy the following dependencies:
  ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ
    /home/mortennk/koding/spread/install/tests/some/dir/test.sh

Add archive but not hint:
CAUGHT: Don't know how to satisfy the following dependencies:
  ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ
    /home/mortennk/koding/spread/install/tests/some/dir/test.sh

Both archive and hint, but no loadable dir:
CAUGHT: Don't know how to satisfy the following dependencies:
  ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ
    /home/mortennk/koding/spread/install/tests/some/dir/test.sh

WITH loadable dir:
CAUGHT: Don't know how to satisfy the following dependencies:
  bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ

Full working case:
Got 2 actions:
bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ : COPY from /home/mortennk/koding/spread/install/tests/testsh.zip
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : ZIP1
  <= bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
  => /home/mortennk/koding/spread/install/tests/some/dir/test.sh

Implicit hinting through addDir:
Got 2 actions:
bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ : COPY from /home/mortennk/koding/spread/install/tests/testsh.zip
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : ZIP1
  <= bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
  => /home/mortennk/koding/spread/install/tests/out/test.sh

Three files from different sources, one missing:
CAUGHT: Don't know how to satisfy the following dependencies:
  SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF
    /home/mortennk/koding/spread/install/tests/out/dir/world.out

Three files from different sources, all present:
Got 4 actions:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : COPY from /home/mortennk/koding/spread/install/tests/hello.txt
  => /home/mortennk/koding/spread/install/tests/out/hello.out
SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF : URL SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF http://world-url.com/
  => /home/mortennk/koding/spread/install/tests/out/dir/world.out
bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ : COPY from /home/mortennk/koding/spread/install/tests/testsh.zip
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : ZIP1
  <= bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
  => /home/mortennk/koding/spread/install/tests/out/test.sh

Overwriting existing file with wrong data:
CAUGHT: Don't know how to satisfy the following dependencies:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
    /home/mortennk/koding/spread/install/tests/test.sh

Overwriting existing file with right data:
Got 1 actions:
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : COPY from /home/mortennk/koding/spread/install/tests/test.sh

Unzipping two zips into one dir:
Got 5 actions:
HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ : COPY from /home/mortennk/koding/spread/install/tests/testdir.zip
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : TESTDIR ZIP
  <= HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ
  => /home/mortennk/koding/spread/install/tests/out/hello.out
SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF : TESTDIR ZIP
  <= HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ
  => /home/mortennk/koding/spread/install/tests/out/dir/world.out
bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ : URL bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ zip1URL
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : TESTSH ZIP
  <= bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
  => /home/mortennk/koding/spread/install/tests/out/test.sh

Hinting to two zips (no actual deps):
Got 0 actions:

Adding stand-alone files from two zips:
Got 5 actions:
HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ : COPY from /home/mortennk/koding/spread/install/tests/testdir.zip
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : TESTDIR ZIP
  <= HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ
  => /home/mortennk/koding/spread/install/tests/file_hello
  => /home/mortennk/koding/spread/install/tests/file_hello_again
SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF : TESTDIR ZIP
  <= HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ
  => /home/mortennk/koding/spread/install/tests/file_world
bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ : URL bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ zip1URL
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : TESTSH ZIP
  <= bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
  => /home/mortennk/koding/spread/install/tests/file_testsh

Zip in zip, hinting at inner zip:
CAUGHT: Don't know how to satisfy the following dependencies:
  bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ

Zip in zip, hinting at outer zip:
CAUGHT: Don't know how to satisfy the following dependencies:
  ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ
    /home/mortennk/koding/spread/install/tests/testsh_output

Zip in zip, hinting at both zips:
Got 3 actions:
RKEFhXsBxTYXmQt4oMxU_RQk_t7F0Oj8RMGALSJKh6UfAg : COPY from /home/mortennk/koding/spread/install/tests/zipzip.zip
bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ : ZIPZIP
  <= RKEFhXsBxTYXmQt4oMxU_RQk_t7F0Oj8RMGALSJKh6UfAg
ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ : ZIP
  <= bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
  => /home/mortennk/koding/spread/install/tests/testsh_output

All URLs present:
Got 1 actions:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : URL LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF 4 url 4
  => /home/mortennk/koding/spread/install/tests/output.hello

Disabled url 4:
Got 1 actions:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : URL LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF 3 10.000000 url 3 10
  => /home/mortennk/koding/spread/install/tests/output.hello

Disabled the other url also:
Got 1 actions:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : URL LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF url default
  => /home/mortennk/koding/spread/install/tests/output.hello

Disabled everything:
CAUGHT: Don't know how to satisfy the following dependencies:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
    /home/mortennk/koding/spread/install/tests/output.hello

